### PR TITLE
add deephash configure files and dch algorithm

### PR DESCRIPTION
**Pull request by Intsigstephon**
1) configure file:  dch.yaml, dshsd.yaml, lcdsh.yaml
2) loss: dchloss
3) data: CustomizedCifar10
4) optimizer: SGD

### CODE REVIEW

File: ppcls/configs/DeepHash/DCH.yaml [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+  name: RecModel
+  infer_output_key: features
+  infer_add_softmax: False
+
+  Backbone:
+    name: AlexNet
+    pretrained: True
+    class_num: 48
+
+# loss function config for traing/eval process
```
Comment (weisy11) [inline ppcls/configs/DeepHash/DCH.yaml]: typo

File: ppcls/data/__init__.py [hunk 2]
```diff
@@ -28,6 +28,7 @@
 from ppcls.data.dataloader.logo_dataset import LogoDataset
 from ppcls.data.dataloader.icartoon_dataset import ICartoonDataset
 from ppcls.data.dataloader.mix_dataset import MixDataset
+from ppcls.data.dataloader.customized_cifar10 import CustomizedCifar10
```
Comment (weisy11) [inline ppcls/data/__init__.py]: 这个dataset的名字可以再明确些，customized没有信息量

File: ppcls/data/dataloader/customized_cifar10.py [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+from PIL import Image
+
+
+class CustomizedCifar10(Cifar10):
+    def __init__(self,
+                 data_file=None,
+                 mode='train',
+                 download=True,
+                 backend=None):
+        assert mode.lower() in ['train', 'test', 'train', 'test'], \
```
Comment (weisy11) [inline ppcls/data/dataloader/customized_cifar10.py]: 为啥有两遍

File: ppcls/data/dataloader/customized_cifar10.py [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+            "mode should be 'train10', 'test10', 'train100' or 'test100', but got {}".format(mode)
```
Comment (weisy11) [inline ppcls/data/dataloader/customized_cifar10.py]: 这个注释跟上面的选项对不上

File: ppcls/data/dataloader/customized_cifar10.py [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+
+        self._init_url_md5_flag()
+
+        self.data_file = data_file
+        if self.data_file is None:
+            assert download, "data_file is not set and downloading automatically is disabled"
+            self.data_file = _check_exists_and_download(
+                data_file, self.data_url, self.data_md5, 'cifar', download)
+
+        self.transform = transforms.Compose([
```
Comment (weisy11) [inline ppcls/data/dataloader/customized_cifar10.py]: transform为什么不通过配置文件来配置

File: ppcls/data/dataloader/customized_cifar10.py [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+            transforms.Resize(224), transforms.ToTensor(),
+            transforms.Normalize([0.485, 0.456, 0.406], [0.229, 0.224, 0.225])
+        ])
+
+        self._load_data()
+        self.dtype = paddle.get_default_dtype()
+
+    def __getitem__(self, index):
+        img, target = self.data[index]
+        img = np.reshape(img, [3, 32, 32])
```
Comment (weisy11) [inline ppcls/data/dataloader/customized_cifar10.py]: 这里的transform为什么是直接写的

File: ppcls/loss/deephashloss.py [hunk 7]
```diff
@@ -37,9 +39,9 @@ def forward(self, input, label):
                           axis=2)
 
         # label to ont-hot
-        label = paddle.flatten(label)
         n_class = logits.shape[1]
-        label = paddle.nn.functional.one_hot(label, n_class).astype("float32")
+        label = paddle.nn.functional.one_hot(
+            label, n_class).astype("float32").squeeze()
```
Comment (weisy11) [inline ppcls/loss/deephashloss.py]: 这个操作顺序跟原来不一样了，为什么要这样改，结果有无变化？

File: ppcls/loss/deephashloss.py [hunk 8]
```diff
@@ -73,11 +76,11 @@ def __init__(self, n_class, _lambda):
...
-        # label to ont-hot
-        label = paddle.flatten(label)
-        label = paddle.nn.functional.one_hot(label,  self.n_class).astype("float32")
-        
-        s = 2 * (paddle.matmul(label, label, transpose_y=True) > 0).astype("float32") - 1
+        label = paddle.nn.functional.one_hot(
+            label, self.n_class).astype("float32").squeeze()
+
+        s = 2 * (paddle.matmul(
+            label, label, transpose_y=True) > 0).astype("float32") - 1
```
Comment (weisy11) [inline ppcls/loss/deephashloss.py]: 表达式太长了，拆一下

File: ppcls/loss/deephashloss.py [hunk 9]
```diff
@@ -90,3 +93,58 @@ def forward(self, input, label):
...
+
+class DCHLoss(paddle.nn.Layer):
+    """
+    # paper [Deep Cauchy Hashing for Hamming Space Retrieval]
+    URL:(http://ise.thss.tsinghua.edu.cn/~mlong/doc/deep-cauchy-hashing-cvpr18.pdf)
+
+    # [DCH] epoch:150, bit:48, dataset:cifar10-1, MAP:0.768, Best MAP: 0.810
+    # [DCH] epoch:150, bit:48, dataset:coco, MAP:0.665, Best MAP: 0.670
+    # [DCH] epoch:150, bit:48, dataset:imagenet, MAP:0.586, Best MAP: 0.586
+    # [DCH] epoch:150, bit:48, dataset:nuswide_21, MAP:0.778, Best MAP: 0.794
```
Comment (weisy11) [inline ppcls/loss/deephashloss.py]: 这个注释是设么含义？为什么写在这里

File: ppcls/loss/deephashloss.py [hunk 10]
```diff
@@ -90,3 +93,58 @@ def forward(self, input, label):
...
+    """
+
+    def __init__(self, gamma, _lambda, n_class):
+        super(DCHLoss, self).__init__()
+        self.gamma = gamma
+        self._lambda = _lambda
+        self.n_class = n_class
+
+    def d(self, hi, hj):
```
Comment (weisy11) [inline ppcls/loss/deephashloss.py]: d hi hj 是什么含义？变量名太随意了

File: ppcls/optimizer/optimizer.py [hunk 11]
```diff
@@ -22,6 +22,53 @@
 from ppcls.utils import logger
 
 
+class SGD(object):
```
Comment (weisy11) [inline ppcls/optimizer/optimizer.py:25]: 为什么要加一个SGD？paddle的SGD有diff吗？

File: ppcls/optimizer/optimizer.py [hunk 12]
```diff
@@ -22,6 +22,53 @@
...
+    def __call__(self, model_list):
+        # model_list is None in static graph
+        parameters = sum([m.parameters() for m in model_list],
+                         []) if model_list else None
+        opt = optim.SGD(learning_rate=self.learning_rate,
+                        parameters=parameters,
+                        weight_decay=self.weight_decay,
+                        grad_clip=self.grad_clip,
+                        name=self.name)
+        return opt
```
Comment (weisy11) [inline ppcls/optimizer/optimizer.py:67]: 从实现来看，这个类是多余的。现有的optimizer部分实现有什么问题呢？

**Review by weisy11 — CHANGES_REQUESTED** [review]

File: ppcls/configs/DeepHash/DCH.yaml (hunk 1 above)
Comment (Intsigstephon) [reply ppcls/configs/DeepHash/DCH.yaml]: 这个是啥意思？

File: ppcls/data/__init__.py (hunk 2 above)
Comment (Intsigstephon) [reply ppcls/data/__init__.py]: 这个数据类是基于paddle已有数据集，修改了接口和transform部分，所以是定制的Cifar10；有更好的名字推荐吗？

File: ppcls/data/dataloader/customized_cifar10.py (hunk 3 above)
Comment (Intsigstephon) [reply ppcls/data/dataloader/customized_cifar10.py]: 这个是从源码复制过来的，确实得改下

File: ppcls/data/dataloader/customized_cifar10.py (hunk 4 above)
Comment (Intsigstephon) [reply ppcls/data/dataloader/customized_cifar10.py]: OK, 这儿得改；paddle的源码也有类似问题

File: ppcls/data/dataloader/customized_cifar10.py (hunk 5 above)
Comment (Intsigstephon) [reply ppcls/data/dataloader/customized_cifar10.py]: 区别于Common_dataset的接口；和paddle内置的数据集接口有较大差异；transform不太好配置，或者得设计另外一种配置格式

File: ppcls/data/dataloader/customized_cifar10.py (hunk 6 above)
Comment (Intsigstephon) [reply ppcls/data/dataloader/customized_cifar10.py]: 可以优化下

File: ppcls/loss/deephashloss.py (hunk 8 above)
Comment (Intsigstephon) [reply ppcls/loss/deephashloss.py]: 长不是问题

File: ppcls/loss/deephashloss.py (hunk 9 above)
Comment (Intsigstephon) [reply ppcls/loss/deephashloss.py]: OK. 这块注释可以再优化下

File: ppcls/loss/deephashloss.py (hunk 10 above)
Comment (Intsigstephon) [reply ppcls/loss/deephashloss.py]: OK

File: ppcls/data/dataloader/customized_cifar10.py (hunk 5 above)
Comment (weisy11) [reply ppcls/data/dataloader/customized_cifar10.py]: 这里用内置的Cifar数据集有什么优势吗？

File: ppcls/loss/deephashloss.py (hunk 7 above)
Comment (Intsigstephon) [reply ppcls/loss/deephashloss.py]: flatte操作和squeeze()是等价的，对结果不会有什么变化

File: ppcls/optimizer/optimizer.py (hunk 12 above)
Comment (Intsigstephon) [reply ppcls/optimizer/optimizer.py:67]: 不多余。对paramters部分做了修改，和原来的不一样

**Review by weisy11 — APPROVED** [review]
LGTM
